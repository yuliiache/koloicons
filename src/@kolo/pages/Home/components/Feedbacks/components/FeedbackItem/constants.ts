export const FEEDBACK_TEXT_TEST_ID = 'feedbackText';
