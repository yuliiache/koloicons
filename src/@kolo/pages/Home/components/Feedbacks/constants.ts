export const FEEDBACKS_KEY = 'feedbacks';
