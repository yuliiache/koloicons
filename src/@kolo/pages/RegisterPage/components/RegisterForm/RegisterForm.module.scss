@import "@kolo/design";

.registerForm {
  margin-bottom: 24px;
}

.checkbox {
  &Wrapper {
    display: flex;
    justify-content: flex-start;

    &:not(:last-of-type) {
      margin-bottom: 8px;
    }

    &:last-of-type {
      margin-bottom: 16px;
    }
  }

  &Text {
    @include main-text(0.7rem, 1.43, null);
    color: $secFontColor;

    &.textDesktop {
      display: block;

      @media (max-width: $mobileWidth) {
        display: none;
      }
    }

    &.textMobile {
      display: none;

      @media (max-width: $mobileWidth) {
        display: block;
      }
    }

    & > a {
      @include main-text(0.7rem, 1.43, null);
      color: $secFontColor;
    }
  }
}
