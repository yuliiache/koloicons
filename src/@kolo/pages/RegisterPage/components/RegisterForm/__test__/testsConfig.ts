import {InputType} from '@kolo/uiKit/Input/InputType';

import {Props as RegisterField} from '../components/FormInputField/FormInputField';

export const REGISTER_FIELDS: readonly RegisterField[] = [
  {fieldName: 'firstName', label: 'First name', type: InputType.TEXT},
  {fieldName: 'lastName', label: 'Last name', type: InputType.TEXT},
  {fieldName: 'email', label: 'Email', type: InputType.TEXT},
  {fieldName: 'password', label: 'Password', type: InputType.PASSWORD},
];
