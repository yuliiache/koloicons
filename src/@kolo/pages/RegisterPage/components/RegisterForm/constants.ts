export const REGISTER_FORM_TEST_ID = 'registerFormComponent';

export const enum CheckboxDataTestId {
  SUBSCRIBE_AGREEMENT_CHECKBOX = 'subscribeAgreementCheckbox',
  TERMS_AGREEMENT_CHECKBOX = 'termsAgreementCheckbox',
}
