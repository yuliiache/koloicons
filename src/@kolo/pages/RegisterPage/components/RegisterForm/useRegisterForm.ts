import {FormApi} from 'final-form';

import {RegisterFormValues} from './types/RegisterFormValues';

interface RegisterFormHookReturn {
  initialValues: RegisterFormValues;
  onFormSubmit: (values: RegisterFormValues, form: FormApi<RegisterFormValues, RegisterFormValues>) => void;
}

const initialValues: RegisterFormValues = {
  firstName: '',
  lastName: '',
  email: '',
  password: '',
  subscribeAgreement: true,
  termsAgreement: false,
};

export const useRegisterForm = (onSubmit: (values: RegisterFormValues) => void): RegisterFormHookReturn => {
  const onFormSubmit = (values: RegisterFormValues, form: FormApi<RegisterFormValues, RegisterFormValues>): void => {
    onSubmit(values);
    form.reset();

    /* reset field's flags (e.g. modified, touched, etc.) to their initial state
    to display errors correctly */
    form.batch(() => {
      Object.keys(initialValues).forEach((fieldName: string) => {
        form.resetFieldState(fieldName as keyof RegisterFormValues);
      });
    });
  };

  return {initialValues, onFormSubmit};
};
