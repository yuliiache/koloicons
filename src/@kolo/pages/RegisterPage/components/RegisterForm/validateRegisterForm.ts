import {
  validateCheckbox,
  validateEmail,
  validateFirstName,
  validateLastName,
  validatePassword,
} from '@kolo/services/validators/commonValidators';

import {RegisterFormValues} from './types/RegisterFormValues';

interface ValidationErrors {
  firstName?: string;
  lastName?: string;
  email?: string;
  password?: string;
  termsAgreement?: boolean;
}

export const validateRegisterForm = (values: RegisterFormValues): ValidationErrors => {
  const errors: ValidationErrors = {};
  const {firstName, lastName, email, password, termsAgreement} = values;

  errors.firstName = validateFirstName(firstName);
  errors.lastName = validateLastName(lastName);
  errors.email = validateEmail(email);
  errors.password = validatePassword(password);
  errors.termsAgreement = validateCheckbox(termsAgreement);

  return errors;
};
