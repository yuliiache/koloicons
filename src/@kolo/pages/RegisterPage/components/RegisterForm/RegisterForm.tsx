import Button from '@kolo/uiKit/Button/Button';
import {ButtonSize, ButtonType} from '@kolo/uiKit/Button/constants';
import {InputType} from '@kolo/uiKit/Input/InputType';
import Link from '@kolo/uiKit/Link/Link';
import classNames from 'classnames';
import {AppRoute} from 'constants/AppRoute';
import {FC} from 'react';
import {Form, FormRenderProps} from 'react-final-form';

import FormCheckboxField from './components/FormCheckboxField/FormCheckboxField';
import FormInputField from './components/FormInputField/FormInputField';
import {CheckboxDataTestId, REGISTER_FORM_TEST_ID} from './constants';
import styles from './RegisterForm.module.scss';
import {RegisterFormValues} from './types/RegisterFormValues';
import {useRegisterForm} from './useRegisterForm';
import {validateRegisterForm} from './validateRegisterForm';

interface Props {
  onSubmit: (values: RegisterFormValues) => void;
}

const RegisterForm: FC<Props> = ({onSubmit}) => {
  const {initialValues, onFormSubmit} = useRegisterForm(onSubmit);

  return (
    <div data-testid={REGISTER_FORM_TEST_ID}>
      <Form
        onSubmit={onFormSubmit}
        initialValues={initialValues}
        subscription={{submitError: true}}
        validate={validateRegisterForm}
        render={(props: FormRenderProps<RegisterFormValues>) => (
          <form
            className={styles.registerForm}
            onSubmit={props.handleSubmit}
          >
            <FormInputField
              fieldName={'firstName'}
              label={'First name'}
              type={InputType.TEXT}
            />
            <FormInputField
              fieldName={'lastName'}
              label={'Last name'}
              type={InputType.TEXT}
            />
            <FormInputField
              fieldName={'email'}
              label={'Email'}
              type={InputType.TEXT}
            />
            <FormInputField
              fieldName={'password'}
              label={'Password'}
              type={InputType.PASSWORD}
            />

            <div className={styles.checkboxWrapper}>
              <FormCheckboxField
                dataTestId={CheckboxDataTestId.SUBSCRIBE_AGREEMENT_CHECKBOX}
                fieldName="subscribeAgreement"
              />

              <p className={styles.checkboxText}>
                Agree to receive emails from koloicons with discounts or special offers.
              </p>
            </div>

            <div className={styles.checkboxWrapper}>
              <FormCheckboxField
                dataTestId={CheckboxDataTestId.TERMS_AGREEMENT_CHECKBOX}
                fieldName="termsAgreement"
              />

              <p className={classNames(styles.checkboxText, styles.textDesktop)}>
                Agree to KOLO Icons Terms of Use and Privacy Policy.
              </p>
              <p className={classNames(styles.checkboxText, styles.textMobile)}>
                Agree to KOLO Icons{' '}
                <Link url={AppRoute.TERMS_OF_SERVICE}>
                  <span>Terms of Use</span>
                </Link>{' '}
                and{' '}
                <Link url={AppRoute.PRIVACY_POLICY}>
                  <span>Privacy Policy</span>
                </Link>
                .
              </p>
            </div>

            <Button
              size={ButtonSize.FULL_WIDTH}
              text="Sign Up"
              type={ButtonType.SUBMIT}
            />
          </form>
        )}
      />
    </div>
  );
};

export default RegisterForm;
