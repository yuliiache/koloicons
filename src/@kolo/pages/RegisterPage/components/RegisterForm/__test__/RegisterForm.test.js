import {BrowserRouter, MemoryRouter} from 'react-router-dom';
import {cleanup, fireEvent, render, screen} from '@testing-library/react';
import {AppRoute} from 'constants/AppRoute';

import {CheckboxDataTestId} from '../constants';
import RegisterForm from '../RegisterForm';

import {REGISTER_FIELDS} from './testsConfig';

describe('RegisterForm component', () => {
  const onSubmitMock = jest.fn();
  const renderForm = () => render(<RegisterForm onSubmit={onSubmitMock} />, {wrapper: MemoryRouter});

  it('should render snapshot successfully', () => {
    const {asFragment} = renderForm();

    expect(asFragment()).toMatchSnapshot();
  });

  it('should navigate to "Terms of service" page when user click the "Terms of Use" link', () => {
    render(<RegisterForm onSubmit={onSubmitMock} />, {wrapper: BrowserRouter});
    const link = screen.getByRole('link', {name: 'Terms of Use'});

    fireEvent.click(link);

    expect(window.location.pathname).toEqual(AppRoute.TERMS_OF_SERVICE);
  });

  it('should navigate to "Privacy Policy" page when user click the "Privacy Policy" link', () => {
    render(<RegisterForm onSubmit={onSubmitMock} />, {wrapper: BrowserRouter});
    const link = screen.getByRole('link', {name: 'Privacy Policy'});

    fireEvent.click(link);

    expect(window.location.pathname).toEqual(AppRoute.PRIVACY_POLICY);
  });

  it('form with the correct button', () => {
    renderForm();
    const submitButton = screen.getByRole('button');

    expect(submitButton).toHaveTextContent('Sign Up');
  });
});

describe('RegisterForm component', () => {
  afterEach(cleanup);
  const mockValidInputValues = {
    firstName: 'Valid First Name',
    lastName: 'Valid Last Name',
    email: 'valid.Email@example.com',
    password: 'validPassword1',
  };
  const onSubmitMock = jest.fn();
  const renderForm = () => render(<RegisterForm onSubmit={onSubmitMock} />, {wrapper: MemoryRouter});
  const submitForm = () => {
    fireEvent.submit(screen.getByRole('button', {name: 'Sign Up'}));
  };

  it('should call onSubmit function', () => {
    renderForm();
    const mockValidFormValues = {
      ...mockValidInputValues,
      subscribeAgreement: true,
      termsAgreement: true,
    };

    const firstNameInput = screen.getByLabelText('First name');
    const lastNameInput = screen.getByLabelText('Last name');
    const emailInput = screen.getByLabelText('Email');
    const passwordInput = screen.getByLabelText('Password');
    const checkboxWrapper = screen.getByTestId(CheckboxDataTestId.TERMS_AGREEMENT_CHECKBOX);
    const termsAgreementCheckbox = checkboxWrapper.querySelector('input[type="checkbox"]');

    fireEvent.change(firstNameInput, {target: {value: mockValidFormValues.firstName}});
    fireEvent.change(lastNameInput, {target: {value: mockValidFormValues.lastName}});
    fireEvent.change(emailInput, {target: {value: mockValidFormValues.email}});
    fireEvent.change(passwordInput, {target: {value: mockValidFormValues.password}});
    fireEvent.click(termsAgreementCheckbox);

    fireEvent.click(screen.getByText('Sign Up'));

    expect(onSubmitMock).toHaveBeenCalledWith(mockValidFormValues);
  });

  describe('should show validation error on submit', () => {
    afterEach(cleanup);

    REGISTER_FIELDS.forEach(({label}) => {
      it(`for the ${label} input`, () => {
        renderForm();
        submitForm();

        expect(screen.getByText(`Please, fill in ${label.toLowerCase()}`)).toBeInTheDocument();
      });
    });

    it('for terms of use agreement checkbox', () => {
      renderForm();
      submitForm();

      expect(screen.getByText('Required')).toBeInTheDocument();
    });

    it('and render snapshot successfully', () => {
      const {asFragment} = renderForm();
      submitForm();

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('should not show validation error on submit', () => {
    beforeEach(() => {
      renderForm();
    });
    afterEach(cleanup);

    REGISTER_FIELDS.forEach(({label}) => {
      Object.keys(mockValidInputValues).forEach((fieldName) => {
        it(`for the ${label} input`, () => {
          const errorMessage = `Please, fill in ${label.toLowerCase()}`;
          const input = screen.getByLabelText(`${label}`);
          const inputValue = mockValidInputValues[fieldName];

          fireEvent.change(input, {target: {value: inputValue}});

          submitForm();

          expect(input).toHaveValue(inputValue);
          expect(screen.queryByText(errorMessage)).not.toBeInTheDocument();
        });
      });
    });

    it('for the required checkbox', () => {
      const errorMessage = 'Required';

      const checkboxWrapper = screen.getByTestId(CheckboxDataTestId.TERMS_AGREEMENT_CHECKBOX);
      const termsAgreementCheckbox = checkboxWrapper.querySelector('input[type="checkbox"]');

      fireEvent.click(termsAgreementCheckbox);

      submitForm();

      expect(termsAgreementCheckbox).toBeChecked();
      expect(screen.queryByText(errorMessage)).not.toBeInTheDocument();
    });
  });
});
