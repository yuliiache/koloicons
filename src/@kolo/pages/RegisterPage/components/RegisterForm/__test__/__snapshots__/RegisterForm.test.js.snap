// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterForm component should render snapshot successfully 1`] = `
<DocumentFragment>
  <div
    data-testid="registerFormComponent"
  >
    <form
      class="registerForm"
    >
      <div
        class="inputGroup"
      >
        <label
          class="inputGroupLabel"
          for="firstName"
        >
          First name
        </label>
        <input
          class="inputGroupInput"
          id="firstName"
          type="text"
          value=""
        />
      </div>
      <div
        class="inputGroup"
      >
        <label
          class="inputGroupLabel"
          for="lastName"
        >
          Last name
        </label>
        <input
          class="inputGroupInput"
          id="lastName"
          type="text"
          value=""
        />
      </div>
      <div
        class="inputGroup"
      >
        <label
          class="inputGroupLabel"
          for="email"
        >
          Email
        </label>
        <input
          class="inputGroupInput"
          id="email"
          type="text"
          value=""
        />
      </div>
      <div
        class="inputGroup"
      >
        <label
          class="inputGroupLabel"
          for="password"
        >
          Password
        </label>
        <input
          class="inputGroupInput"
          id="password"
          type="password"
          value=""
        />
      </div>
      <div
        class="checkboxWrapper"
      >
        <div
          data-testid="subscribeAgreementCheckbox"
        >
          <label
            class="checkbox"
          >
            <input
              checked=""
              class="checkbox__input"
              id="subscribeAgreement"
              type="checkbox"
            />
            <span
              class="checkbox__indicator"
            />
          </label>
        </div>
        <p
          class="checkboxText"
        >
          Agree to receive emails from koloicons with discounts or special offers.
        </p>
      </div>
      <div
        class="checkboxWrapper"
      >
        <div
          data-testid="termsAgreementCheckbox"
        >
          <label
            class="checkbox"
          >
            <input
              class="checkbox__input"
              id="termsAgreement"
              type="checkbox"
            />
            <span
              class="checkbox__indicator"
            />
          </label>
        </div>
        <p
          class="checkboxText textDesktop"
        >
          Agree to KOLO Icons Terms of Use and Privacy Policy.
        </p>
        <p
          class="checkboxText textMobile"
        >
          Agree to KOLO Icons 
          <a
            class="link link--default"
            data-testid="internalLinkComponent"
            href="/terms-of-service"
          >
            <span>
              Terms of Use
            </span>
          </a>
           and 
          <a
            class="link link--default"
            data-testid="internalLinkComponent"
            href="/privacy-policy-docs"
          >
            <span>
              Privacy Policy
            </span>
          </a>
          .
        </p>
      </div>
      <button
        class="button button-full-width button-primary button-center"
        data-testid="buttonComponent"
        type="submit"
      >
        <span
          data-testid="buttonComponentText"
        >
          Sign Up
        </span>
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`RegisterForm component should show validation error on submit and render snapshot successfully 1`] = `
<DocumentFragment>
  <div
    data-testid="registerFormComponent"
  >
    <form
      class="registerForm"
    >
      <div
        class="inputGroup"
      >
        <label
          class="inputGroupLabel"
          for="firstName"
        >
          First name
        </label>
        <input
          class="inputGroupInput invalid"
          id="firstName"
          type="text"
          value=""
        />
        <span
          class="inputGroupError"
        >
          Please, fill in first name
        </span>
      </div>
      <div
        class="inputGroup"
      >
        <label
          class="inputGroupLabel"
          for="lastName"
        >
          Last name
        </label>
        <input
          class="inputGroupInput invalid"
          id="lastName"
          type="text"
          value=""
        />
        <span
          class="inputGroupError"
        >
          Please, fill in last name
        </span>
      </div>
      <div
        class="inputGroup"
      >
        <label
          class="inputGroupLabel"
          for="email"
        >
          Email
        </label>
        <input
          class="inputGroupInput invalid"
          id="email"
          type="text"
          value=""
        />
        <span
          class="inputGroupError"
        >
          Please, fill in email
        </span>
      </div>
      <div
        class="inputGroup"
      >
        <label
          class="inputGroupLabel"
          for="password"
        >
          Password
        </label>
        <input
          class="inputGroupInput invalid"
          id="password"
          type="password"
          value=""
        />
        <span
          class="inputGroupError"
        >
          Please, fill in password
        </span>
      </div>
      <div
        class="checkboxWrapper"
      >
        <div
          data-testid="subscribeAgreementCheckbox"
        >
          <label
            class="checkbox"
          >
            <input
              checked=""
              class="checkbox__input"
              id="subscribeAgreement"
              type="checkbox"
            />
            <span
              class="checkbox__indicator"
            />
          </label>
        </div>
        <p
          class="checkboxText"
        >
          Agree to receive emails from koloicons with discounts or special offers.
        </p>
      </div>
      <div
        class="checkboxWrapper"
      >
        <div
          data-testid="termsAgreementCheckbox"
        >
          <label
            class="checkbox"
          >
            <input
              class="checkbox__input"
              id="termsAgreement"
              type="checkbox"
            />
            <span
              class="checkbox__indicator checkbox__indicator--error"
            />
            <span
              class="checkbox__error"
            >
              Required
            </span>
          </label>
        </div>
        <p
          class="checkboxText textDesktop"
        >
          Agree to KOLO Icons Terms of Use and Privacy Policy.
        </p>
        <p
          class="checkboxText textMobile"
        >
          Agree to KOLO Icons 
          <a
            class="link link--default"
            data-testid="internalLinkComponent"
            href="/terms-of-service"
          >
            <span>
              Terms of Use
            </span>
          </a>
           and 
          <a
            class="link link--default"
            data-testid="internalLinkComponent"
            href="/privacy-policy-docs"
          >
            <span>
              Privacy Policy
            </span>
          </a>
          .
        </p>
      </div>
      <button
        class="button button-full-width button-primary button-center"
        data-testid="buttonComponent"
        type="submit"
      >
        <span
          data-testid="buttonComponentText"
        >
          Sign Up
        </span>
      </button>
    </form>
  </div>
</DocumentFragment>
`;
